src/Coordinate.py

```diff
@@ -1,13 +1,22 @@
-class Coordinate(tuple):
+from __future__ import annotations
 
-    def __new__(cls, *args):
+from typing import Any
+
+
+class Coordinate(tuple[int, int]):
```

Comment:
I think it would be an idea to let this inherit from `typing.NamedTuple` instead from a generic `tuple[int, int]`

Reply:
Potentially, but this also seems to interfere with the existing code too much. I just checked what `mypy` would say if I do this, and it looks there are a few more changes required. I would leave inheriting from `tuple` there as it was here already, I only added `int` types here.

Mypy if I swap `tuple` with `NamedTuple` in line 6:
```
~/Projects/command-line-chess (add-mypy ✗) mypy --show-error-codes --pretty --strict .
src/Coordinate.py:8: error: Cannot overwrite NamedTuple attribute "__new__"  [misc]
        def __new__(cls, *args: int) -> Coordinate:
        ^
src/Coordinate.py:12: error: Incompatible return value type (got "Tuple[Type[Coordinate], Tuple[<nothing>, ...]]", expected "NamedTuple")  [return-value]
            return (self.__class__, tuple(self))
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/Coordinate.py:17: error: Tuple index out of range  [misc]
            return Coordinate(self[0] + other[0], self[1] + other[1])
                              ^~~~~~~
src/Coordinate.py:22: error: Tuple index out of range  [misc]
            return Coordinate(self[0] - other[0], self[1] - other[1])
                              ^~~~~~~
src/Pawn.py:39: error: Tuple index out of range  [misc]
                    col = advanceOnePosition[1]
                          ^~~~~~~~~~~~~~~~~~~~~
src/Pawn.py:72: error: Tuple index out of range  [misc]
                        col = newPosition[1]
                              ^~~~~~~~~~~~~~
src/Pawn.py:92: error: Tuple index out of range  [misc]
                posBesidePawn = self.position + C(movement[0], 0)
                                                  ^~~~~~~~~~~
src/Board.py:245: error: Tuple index out of range  [misc]
            return str(piece.position[1] + 1)
                       ^~~~~~~~~~~~~~~~~
src/Board.py:249: error: Tuple index out of range  [misc]
            return str(piece.position[0]).translate(transTable)
                       ^~~~~~~~~~~~~~~~~
src/Board.py:374: error: Tuple index out of range  [misc]
            notation = str(pos[0]).translate(transTable) + str(pos[1]+1)
                           ^~~~~~
src/Board.py:378: error: Returning Any from function declared to return "bool"  [no-any-return]
            return 0 <= pos[0] <= 7 and 0 <= pos[1] <= 7
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/Board.py:378: error: Tuple index out of range  [misc]
            return 0 <= pos[0] <= 7 and 0 <= pos[1] <= 7
                        ^~~~~~
Found 12 errors in 3 files (checked 16 source files)

```

Reply:
This error eventually comes from because you changed it to be a `NamedTuple`, but didn't specify any fields.
```python
# things before ...

class Coordinate(NamedTuple):
    rank: int
    file: int

# things after ...
```
and mypy is happy with that (at least it didn't show any errors like yours to me)

Reply:
This makes sense and fixes most of the errors, but I still have:

```
~/Projects/command-line-chess (add-mypy ✗) mypy --show-error-codes --pretty --strict .
src/Coordinate.py:10: error: Cannot overwrite NamedTuple attribute "__new__"  [misc]
        def __new__(cls, *args: int) -> Coordinate:
        ^
Found 1 error in 1 file (checked 16 source files)
```

Do you have an idea how to fix that too?

Reply:
I think we probably could remove some or even all dunder methods when switching to `NamedTuple`.

Reply:
Yup, only `__add__` and `_sub__` seem to be needed, thanks.